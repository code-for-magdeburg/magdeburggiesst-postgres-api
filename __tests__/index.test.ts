import each from "jest-each";
import fetch from "cross-fetch";
import { test, describe, expect } from "@jest/globals";
import handler from "../api/v3/index";
import getHandler from "../api/v3/get/index";
import { createTestServer } from "../__test-utils/create-test-server";
import { Methods } from "../_utils/routes-listing";

describe("GET/POST/DELETE routes index", () => {
	test.todo("make a test fail with 500");
	test("should list all routes on /get", async () => {
		const { server, url } = await createTestServer({}, getHandler);
		const response = await fetch(url, {
			method: "GET",
			headers: {
				"Content-Type": "application/json",
			},
		});
		server.close();
		const json = await response.json();
		expect(response.status).toBe(200);
		expect(json).toMatchSnapshot();
	});
});
each([
	{
		statusCode: 200,
		description: "",
		method: "GET",
	},
	{
		statusCode: 200,
		description: "",
		method: "POST",
	},
	{
		statusCode: 200,
		description: "",
		method: "DELETE",
	},
]).describe(
	"GET default api route (healthchecks only)",
	({
		statusCode,
		description,
		method,
	}: {
		statusCode: number;
		description: string;
		method: Methods;
	}) => {
		test(`should return ${statusCode} on route "index" ${description}`, async () => {
			const { server, url } = await createTestServer({}, handler);
			const response = await fetch(url, {
				method,
				headers: {
					"Content-Type": "application/json",
				},
			});
			server.close();
			expect(response.status).toBe(statusCode);
		});
	}
);
