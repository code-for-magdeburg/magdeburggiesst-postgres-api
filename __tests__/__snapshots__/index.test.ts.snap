// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET/POST/DELETE routes index should list all routes on /delete 1`] = `
{
  "message": "its working",
  "name": "@code-for-magdeburg/magdeburggiesst-postgres-api",
  "routes": [
    {
      "method": "GET",
      "routes": {
        "adopted": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "limit": {
                "type": "string",
              },
              "offset": {
                "type": "string",
              },
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
            },
            "required": [
              "uuid",
            ],
            "type": "object",
          },
          "url": "get/adopted",
        },
        "byid": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "id": {
                "type": "string",
              },
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
            },
            "required": [
              "id",
            ],
            "type": "object",
          },
          "url": "get/byid",
        },
        "istreeadopted": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "id": {
                "type": "string",
              },
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
            },
            "required": [
              "uuid",
              "id",
            ],
            "type": "object",
          },
          "url": "get/istreeadopted",
        },
        "lastwatered": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "id": {
                "type": "string",
              },
              "limit": {
                "type": "string",
              },
              "offset": {
                "type": "string",
              },
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
            },
            "required": [
              "id",
            ],
            "type": "object",
          },
          "url": "get/lastwatered",
        },
        "treesbyids": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "limit": {
                "type": "string",
              },
              "offset": {
                "type": "string",
              },
              "tree_ids": {
                "type": "string",
              },
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
            },
            "required": [
              "tree_ids",
            ],
            "type": "object",
          },
          "url": "get/treesbyids",
        },
        "wateredandadopted": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "limit": {
                "type": "string",
              },
              "offset": {
                "type": "string",
              },
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
            },
            "required": [],
            "type": "object",
          },
          "url": "get/wateredandadopted",
        },
        "wateredbyuser": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
            },
            "required": [
              "uuid",
            ],
            "type": "object",
          },
          "url": "get/wateredbyuser",
        },
      },
    },
    {
      "method": "POST",
      "routes": {
        "adopt": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "queryType": {
                "type": "string",
              },
              "tree_id": {
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
            },
            "required": [
              "uuid",
              "tree_id",
            ],
            "type": "object",
          },
          "url": "post/adopt",
        },
        "water": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "amount": {
                "type": "number",
              },
              "queryType": {
                "type": "string",
              },
              "timestamp": {
                "type": "string",
              },
              "tree_id": {
                "type": "string",
              },
              "username": {
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
            },
            "required": [
              "uuid",
              "tree_id",
              "username",
              "timestamp",
              "amount",
            ],
            "type": "object",
          },
          "url": "post/water",
        },
      },
    },
    {
      "method": "DELETE",
      "routes": {
        "unadopt": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "queryType": {
                "type": "string",
              },
              "tree_id": {
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
            },
            "required": [
              "uuid",
              "tree_id",
            ],
            "type": "object",
          },
          "url": "delete/unadopt",
        },
        "unwater": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "queryType": {
                "type": "string",
              },
              "tree_id": {
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
              "watering_id": {
                "type": "number",
              },
            },
            "required": [
              "uuid",
              "tree_id",
              "watering_id",
            ],
            "type": "object",
          },
          "url": "delete/unwater",
        },
      },
    },
  ],
}
`;

exports[`GET/POST/DELETE routes index should list all routes on /get 1`] = `
{
  "message": "its working",
  "name": "@code-for-magdeburg/magdeburggiesst-postgres-api",
  "routes": [
    {
      "method": "GET",
      "routes": {
        "adopted": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "limit": {
                "type": "string",
              },
              "offset": {
                "type": "string",
              },
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
            },
            "required": [
              "uuid",
            ],
            "type": "object",
          },
          "url": "get/adopted",
        },
        "byid": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "id": {
                "type": "string",
              },
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
            },
            "required": [
              "id",
            ],
            "type": "object",
          },
          "url": "get/byid",
        },
        "istreeadopted": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "id": {
                "type": "string",
              },
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
            },
            "required": [
              "uuid",
              "id",
            ],
            "type": "object",
          },
          "url": "get/istreeadopted",
        },
        "lastwatered": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "id": {
                "type": "string",
              },
              "limit": {
                "type": "string",
              },
              "offset": {
                "type": "string",
              },
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
            },
            "required": [
              "id",
            ],
            "type": "object",
          },
          "url": "get/lastwatered",
        },
        "treesbyids": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "limit": {
                "type": "string",
              },
              "offset": {
                "type": "string",
              },
              "tree_ids": {
                "type": "string",
              },
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
            },
            "required": [
              "tree_ids",
            ],
            "type": "object",
          },
          "url": "get/treesbyids",
        },
        "wateredandadopted": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "limit": {
                "type": "string",
              },
              "offset": {
                "type": "string",
              },
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
            },
            "required": [],
            "type": "object",
          },
          "url": "get/wateredandadopted",
        },
        "wateredbyuser": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
            },
            "required": [
              "uuid",
            ],
            "type": "object",
          },
          "url": "get/wateredbyuser",
        },
      },
    },
    {
      "method": "POST",
      "routes": {
        "adopt": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "queryType": {
                "type": "string",
              },
              "tree_id": {
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
            },
            "required": [
              "uuid",
              "tree_id",
            ],
            "type": "object",
          },
          "url": "post/adopt",
        },
        "water": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "amount": {
                "type": "number",
              },
              "queryType": {
                "type": "string",
              },
              "timestamp": {
                "type": "string",
              },
              "tree_id": {
                "type": "string",
              },
              "username": {
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
            },
            "required": [
              "uuid",
              "tree_id",
              "username",
              "timestamp",
              "amount",
            ],
            "type": "object",
          },
          "url": "post/water",
        },
      },
    },
    {
      "method": "DELETE",
      "routes": {
        "unadopt": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "queryType": {
                "type": "string",
              },
              "tree_id": {
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
            },
            "required": [
              "uuid",
              "tree_id",
            ],
            "type": "object",
          },
          "url": "delete/unadopt",
        },
        "unwater": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "queryType": {
                "type": "string",
              },
              "tree_id": {
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
              "watering_id": {
                "type": "number",
              },
            },
            "required": [
              "uuid",
              "tree_id",
              "watering_id",
            ],
            "type": "object",
          },
          "url": "delete/unwater",
        },
      },
    },
  ],
}
`;

exports[`GET/POST/DELETE routes index should list all routes on /post 1`] = `
{
  "message": "its working",
  "name": "@code-for-magdeburg/magdeburggiesst-postgres-api",
  "routes": [
    {
      "method": "GET",
      "routes": {
        "adopted": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "limit": {
                "type": "string",
              },
              "offset": {
                "type": "string",
              },
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
            },
            "required": [
              "uuid",
            ],
            "type": "object",
          },
          "url": "get/adopted",
        },
        "byid": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "id": {
                "type": "string",
              },
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
            },
            "required": [
              "id",
            ],
            "type": "object",
          },
          "url": "get/byid",
        },
        "istreeadopted": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "id": {
                "type": "string",
              },
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
            },
            "required": [
              "uuid",
              "id",
            ],
            "type": "object",
          },
          "url": "get/istreeadopted",
        },
        "lastwatered": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "id": {
                "type": "string",
              },
              "limit": {
                "type": "string",
              },
              "offset": {
                "type": "string",
              },
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
            },
            "required": [
              "id",
            ],
            "type": "object",
          },
          "url": "get/lastwatered",
        },
        "treesbyids": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "limit": {
                "type": "string",
              },
              "offset": {
                "type": "string",
              },
              "tree_ids": {
                "type": "string",
              },
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
            },
            "required": [
              "tree_ids",
            ],
            "type": "object",
          },
          "url": "get/treesbyids",
        },
        "wateredandadopted": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "limit": {
                "type": "string",
              },
              "offset": {
                "type": "string",
              },
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
            },
            "required": [],
            "type": "object",
          },
          "url": "get/wateredandadopted",
        },
        "wateredbyuser": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "type": {
                "description": "The type property is atomaticaly added by dynamic vercel api routes. You should not add it yourself",
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
            },
            "required": [
              "uuid",
            ],
            "type": "object",
          },
          "url": "get/wateredbyuser",
        },
      },
    },
    {
      "method": "POST",
      "routes": {
        "adopt": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "queryType": {
                "type": "string",
              },
              "tree_id": {
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
            },
            "required": [
              "uuid",
              "tree_id",
            ],
            "type": "object",
          },
          "url": "post/adopt",
        },
        "water": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "amount": {
                "type": "number",
              },
              "queryType": {
                "type": "string",
              },
              "timestamp": {
                "type": "string",
              },
              "tree_id": {
                "type": "string",
              },
              "username": {
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
            },
            "required": [
              "uuid",
              "tree_id",
              "username",
              "timestamp",
              "amount",
            ],
            "type": "object",
          },
          "url": "post/water",
        },
      },
    },
    {
      "method": "DELETE",
      "routes": {
        "unadopt": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "queryType": {
                "type": "string",
              },
              "tree_id": {
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
            },
            "required": [
              "uuid",
              "tree_id",
            ],
            "type": "object",
          },
          "url": "delete/unadopt",
        },
        "unwater": {
          "schema": {
            "additionalProperties": false,
            "properties": {
              "queryType": {
                "type": "string",
              },
              "tree_id": {
                "type": "string",
              },
              "uuid": {
                "type": "string",
              },
              "watering_id": {
                "type": "number",
              },
            },
            "required": [
              "uuid",
              "tree_id",
              "watering_id",
            ],
            "type": "object",
          },
          "url": "delete/unwater",
        },
      },
    },
  ],
}
`;
